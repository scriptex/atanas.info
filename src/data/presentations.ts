export type Presentation = {
	readonly id: string;
	readonly url: string;
	readonly title: string;
	readonly description: string;
};

export const presentations: Presentation[] = [
	{
		id: '5xcqprI17FQ',
		url: 'https://www.youtube.com/embed/5xcqprI17FQ',
		title: 'Do more with Sublime Text',
		description: 'Presentation about optimizing the web development workflow using Sublime Text editor'
	},
	{
		id: 's-klXHLIZyY',
		url: 'https://www.youtube.com/embed/s-klXHLIZyY',
		title: 'Jarvis: Frontend project setup',
		description: 'Presentation about the new and optimized frontend projects setup'
	}
];
