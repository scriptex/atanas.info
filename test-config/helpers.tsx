import { act } from 'react-dom/test-utils';
import type { FC } from 'react';
import { render, RenderResult, waitFor } from '@testing-library/react';

import * as shared from '@scripts/shared';

type ComponentType = FC<Readonly<React.ComponentProps<any>>>;

export const test = async (Component: ComponentType): Promise<RenderResult> => {
	let result!: RenderResult;

	await act(async () => {
		result = await waitFor(() => render(<Component />));
	});

	return result;
};

export const snapshotTest = (Component: ComponentType, buttonSelector?: string): void => {
	const { name } = Component;

	jest.spyOn(shared, 'formatDate');

	(shared.formatDate as jest.Mock).mockImplementation(() => 'Mock date');

	afterAll(() => {
		jest.clearAllMocks();
	});

	describe(`${name} component`, () => {
		it(`Should render the ${name} component`, async () => {
			const { asFragment, container } = await test(Component);

			expect(asFragment()).toMatchSnapshot();

			if (buttonSelector) {
				await act(async () => {
					await container.querySelectorAll<HTMLElement>(buttonSelector).forEach(button => {
						if (!button.hasAttribute('download')) {
							button.click();

							expect(asFragment()).toMatchSnapshot();
						}
					});
				});
			}
		});

		it(`Should unmount the ${name} component`, async () => {
			const { unmount, asFragment } = await test(Component);

			unmount();

			expect(asFragment()).toMatchSnapshot();
		});
	});
};

export const mockAudioContext = (): jest.Mock => {
	const mockConnect = jest.fn();

	const mockgetByteFrequencyData = jest.fn();

	const mockcreateMediaElementSource = jest.fn(() => ({
		connect: mockConnect
	}));

	const mockcreateAnalyser = jest.fn(() => ({
		connect: mockConnect,
		frequencyBinCount: [0, 1, 2],
		getByteFrequencyData: mockgetByteFrequencyData
	}));

	const mockcreateOscillator = jest.fn(() => ({
		channelCount: 2
	}));

	const mockChannelSplitterConnect = jest.fn(n => n);

	const mockcreateChannelSplitter = jest.fn(() => ({
		connect: mockChannelSplitterConnect
	}));

	const mockedAudioContext = jest.fn(() => ({
		createAnalyser: mockcreateAnalyser,
		createOscillator: mockcreateOscillator,
		createChannelSplitter: mockcreateChannelSplitter,
		createMediaElementSource: mockcreateMediaElementSource
	}));

	return mockedAudioContext;
};

export const mockFetch = (value: Record<string, string | boolean>): void => {
	global.fetch = jest.fn().mockImplementation(() =>
		Promise.resolve({
			json: () => Promise.resolve(value)
		})
	);
};
