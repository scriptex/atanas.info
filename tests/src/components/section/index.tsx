import { statsItems } from '@data/projects';
import { snapshotTest } from '@test-config/helpers';
import { presentations } from '@data/presentations';
import { Section, SectionNav, SectionGrid } from '@components';

snapshotTest(
	() => (
		<Section
			id="test"
			title="Test title"
			style={{ color: 'rebeccapurple' }}
			actions={<div />}
			subtitle="Test subtitle"
			className="c-section-test"
			shellClass="c-shell-test"
			additionalElements={<div />}
		>
			<h1>This is a test title</h1>
		</Section>
	),
	'.c-section__actions .c-btn:first-child'
);

snapshotTest(() => (
	<Section
		id="test"
		style={{ color: 'rebeccapurple' }}
		actions={<div />}
		subtitle="Test subtitle"
		hasButton={false}
		className="c-section-test"
		shellClass="c-shell-test"
		additionalElements={<div />}
	>
		<h1>This is a test title</h1>
	</Section>
));

snapshotTest(() => (
	<Section
		id="test"
		style={{ color: 'rebeccapurple' }}
		actions={<div />}
		hasShell={false}
		hasButton={false}
		className="c-section-test"
		shellClass="c-shell-test"
		additionalElements={<div />}
	>
		<h1>This is a test title</h1>
	</Section>
));

snapshotTest(
	() => (
		<Section
			id="test"
			style={{ color: 'rebeccapurple' }}
			actions={<div />}
			hasShell={false}
			hasButton
			className="c-section-test"
			shellClass="c-shell-test"
			additionalElements={<div />}
		>
			<h1>This is a test title</h1>
		</Section>
	),
	'.c-section__actions .c-btn:first-child'
);

snapshotTest(() => <SectionNav name="title" data={presentations} active={0} onClick={jest.fn()} />, '.c-btn');

snapshotTest(() => <SectionGrid data={statsItems} />);

snapshotTest(() => <SectionGrid data={statsItems} linkType="internal" />);

snapshotTest(() => <SectionGrid data={statsItems} linkType="external" />);
