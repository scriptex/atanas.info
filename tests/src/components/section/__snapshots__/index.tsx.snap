// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` component Should render the  component 1`] = `
<DocumentFragment>
  <section
    class="c-section c-section--test c-section-test"
    id="test"
    style="color: rebeccapurple;"
  >
    <div
      class="o-shell c-shell-test"
    >
      <header
        class="c-section__header"
      >
        <div />
        <h2>
          Test title
        </h2>
        <h3>
          Test subtitle
        </h3>
      </header>
      <h1>
        This is a test title
      </h1>
      <div
        class="c-section__actions"
      >
        <button
          class="css-8n53dr-Button"
          type="button"
        >
          Get in touch
        </button>
        <ul
          class="c-section__actions-list"
        >
          <li>
            <button
              class="css-8n53dr-Button"
              type="button"
            >
              Close
            </button>
          </li>
          <li>
            <button
              class="css-8n53dr-Button"
              type="button"
            >
              <svg
                class="c-section__actions-icon"
              >
                <use
                  xlink:href="#svg-email"
                />
              </svg>
              <span>
                Contact me
              </span>
              <strong>
                directly
              </strong>
            </button>
          </li>
          <li>
            <a
              class="css-8n53dr-Button"
              href="https://app.usebraintrust.com/talent/782/"
            >
              <svg
                class="c-section__actions-icon"
              >
                <use
                  xlink:href="#svg-braintrust"
                />
              </svg>
              <span>
                Hire me on
              </span>
              <strong>
                Braintrust
              </strong>
            </a>
          </li>
          <li>
            <a
              class="css-8n53dr-Button"
              href="https://www.toptal.com/resume/atanas-atanasov"
            >
              <svg
                class="c-section__actions-icon"
              >
                <use
                  xlink:href="#svg-toptal"
                />
              </svg>
              <span>
                Hire me on
              </span>
              <strong>
                Toptal
              </strong>
            </a>
          </li>
        </ul>
        <br
          class="visible-xs-block"
        />
        <div />
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[` component Should render the  component 2`] = `
<DocumentFragment>
  <section
    class="c-section c-section--test c-section-test"
    id="test"
    style="color: rebeccapurple;"
  >
    <div
      class="o-shell c-shell-test"
    >
      <header
        class="c-section__header"
      >
        <div />
        <h3>
          Test subtitle
        </h3>
      </header>
      <h1>
        This is a test title
      </h1>
    </div>
  </section>
</DocumentFragment>
`;

exports[` component Should render the  component 3`] = `
<DocumentFragment>
  <section
    class="c-section c-section--test c-section-test"
    id="test"
    style="color: rebeccapurple;"
  >
    <h1>
      This is a test title
    </h1>
  </section>
</DocumentFragment>
`;

exports[` component Should render the  component 4`] = `
<DocumentFragment>
  <section
    class="c-section c-section--test c-section-test"
    id="test"
    style="color: rebeccapurple;"
  >
    <h1>
      This is a test title
    </h1>
    <div
      class="c-section__actions"
    >
      <button
        class="css-8n53dr-Button"
        type="button"
      >
        Get in touch
      </button>
      <ul
        class="c-section__actions-list"
      >
        <li>
          <button
            class="css-8n53dr-Button"
            type="button"
          >
            Close
          </button>
        </li>
        <li>
          <button
            class="css-8n53dr-Button"
            type="button"
          >
            <svg
              class="c-section__actions-icon"
            >
              <use
                xlink:href="#svg-email"
              />
            </svg>
            <span>
              Contact me
            </span>
            <strong>
              directly
            </strong>
          </button>
        </li>
        <li>
          <a
            class="css-8n53dr-Button"
            href="https://app.usebraintrust.com/talent/782/"
          >
            <svg
              class="c-section__actions-icon"
            >
              <use
                xlink:href="#svg-braintrust"
              />
            </svg>
            <span>
              Hire me on
            </span>
            <strong>
              Braintrust
            </strong>
          </a>
        </li>
        <li>
          <a
            class="css-8n53dr-Button"
            href="https://www.toptal.com/resume/atanas-atanasov"
          >
            <svg
              class="c-section__actions-icon"
            >
              <use
                xlink:href="#svg-toptal"
              />
            </svg>
            <span>
              Hire me on
            </span>
            <strong>
              Toptal
            </strong>
          </a>
        </li>
      </ul>
      <br
        class="visible-xs-block"
      />
      <div />
    </div>
  </section>
</DocumentFragment>
`;

exports[` component Should render the  component 5`] = `
<DocumentFragment>
  <nav
    class="c-section__nav"
  >
    <ul>
      <li
        class="current"
      >
        <button
          aria-label="Switch to Do more with Sublime Text"
          class="css-8n53dr-Button"
          type="button"
        >
          Do more with Sublime Text
        </button>
      </li>
      <li>
        <button
          aria-label="Switch to Jarvis: Frontend project setup"
          class="css-8n53dr-Button"
          type="button"
        >
          Jarvis: Frontend project setup
        </button>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;

exports[` component Should render the  component 6`] = `
<DocumentFragment>
  <div
    class="c-section__grid o-grid"
  >
    <div
      class="o-grid__item xs-12 sm-6 md-4 lg-4 xl-4"
    >
      <a
        class="c-article-link fullsize-background"
        href="/stats/github"
        style="background-image: url(/images/temp/github.jpg);"
      >
        <strong>
          Github
        </strong>
      </a>
      <div
        class="css-9g2z7l-Loader"
      >
        <div
          class="css-1o0ykwt-Loader"
        />
      </div>
    </div>
    <div
      class="o-grid__item xs-12 sm-6 md-4 lg-4 xl-4"
    >
      <a
        class="c-article-link fullsize-background"
        href="/stats/gitlab"
        style="background-image: url(/images/temp/gitlab.jpg);"
      >
        <strong>
          Gitlab
        </strong>
      </a>
      <div
        class="css-9g2z7l-Loader"
      >
        <div
          class="css-1o0ykwt-Loader"
        />
      </div>
    </div>
    <div
      class="o-grid__item xs-12 sm-6 md-4 lg-4 xl-4"
    >
      <a
        class="c-article-link fullsize-background"
        href="/stats/npm"
        style="background-image: url(/images/temp/npm.jpg);"
      >
        <strong>
          NPM
        </strong>
      </a>
      <div
        class="css-9g2z7l-Loader"
      >
        <div
          class="css-1o0ykwt-Loader"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[` component Should render the  component 7`] = `
<DocumentFragment>
  <div
    class="c-section__grid o-grid"
  >
    <div
      class="o-grid__item xs-12 sm-6 md-4 lg-4 xl-4"
    >
      <a
        class="c-article-link fullsize-background"
        href="/stats/github"
        style="background-image: url(/images/temp/github.jpg);"
      >
        <strong>
          Github
        </strong>
      </a>
      <div
        class="css-9g2z7l-Loader"
      >
        <div
          class="css-1o0ykwt-Loader"
        />
      </div>
    </div>
    <div
      class="o-grid__item xs-12 sm-6 md-4 lg-4 xl-4"
    >
      <a
        class="c-article-link fullsize-background"
        href="/stats/gitlab"
        style="background-image: url(/images/temp/gitlab.jpg);"
      >
        <strong>
          Gitlab
        </strong>
      </a>
      <div
        class="css-9g2z7l-Loader"
      >
        <div
          class="css-1o0ykwt-Loader"
        />
      </div>
    </div>
    <div
      class="o-grid__item xs-12 sm-6 md-4 lg-4 xl-4"
    >
      <a
        class="c-article-link fullsize-background"
        href="/stats/npm"
        style="background-image: url(/images/temp/npm.jpg);"
      >
        <strong>
          NPM
        </strong>
      </a>
      <div
        class="css-9g2z7l-Loader"
      >
        <div
          class="css-1o0ykwt-Loader"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[` component Should render the  component 8`] = `
<DocumentFragment>
  <div
    class="c-section__grid o-grid"
  >
    <div
      class="o-grid__item xs-12 sm-6 md-4 lg-4 xl-4"
    >
      <a
        class="c-article-link fullsize-background"
        href="/stats/github"
        rel="noopener noreferrer"
        style="background-image: url(/images/temp/github.jpg);"
        target="_blank"
      >
        <strong>
          Github
        </strong>
        <svg
          class="c-svg-external-link"
        >
          <use
            xlink:href="#svg-external-link"
          />
        </svg>
      </a>
      <div
        class="css-9g2z7l-Loader"
      >
        <div
          class="css-1o0ykwt-Loader"
        />
      </div>
    </div>
    <div
      class="o-grid__item xs-12 sm-6 md-4 lg-4 xl-4"
    >
      <a
        class="c-article-link fullsize-background"
        href="/stats/gitlab"
        rel="noopener noreferrer"
        style="background-image: url(/images/temp/gitlab.jpg);"
        target="_blank"
      >
        <strong>
          Gitlab
        </strong>
        <svg
          class="c-svg-external-link"
        >
          <use
            xlink:href="#svg-external-link"
          />
        </svg>
      </a>
      <div
        class="css-9g2z7l-Loader"
      >
        <div
          class="css-1o0ykwt-Loader"
        />
      </div>
    </div>
    <div
      class="o-grid__item xs-12 sm-6 md-4 lg-4 xl-4"
    >
      <a
        class="c-article-link fullsize-background"
        href="/stats/npm"
        rel="noopener noreferrer"
        style="background-image: url(/images/temp/npm.jpg);"
        target="_blank"
      >
        <strong>
          NPM
        </strong>
        <svg
          class="c-svg-external-link"
        >
          <use
            xlink:href="#svg-external-link"
          />
        </svg>
      </a>
      <div
        class="css-9g2z7l-Loader"
      >
        <div
          class="css-1o0ykwt-Loader"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[` component Should unmount the  component 1`] = `<DocumentFragment />`;

exports[` component Should unmount the  component 2`] = `<DocumentFragment />`;

exports[` component Should unmount the  component 3`] = `<DocumentFragment />`;

exports[` component Should unmount the  component 4`] = `<DocumentFragment />`;

exports[` component Should unmount the  component 5`] = `<DocumentFragment />`;

exports[` component Should unmount the  component 6`] = `<DocumentFragment />`;

exports[` component Should unmount the  component 7`] = `<DocumentFragment />`;

exports[` component Should unmount the  component 8`] = `<DocumentFragment />`;
